src/vs/base/browser/ui/selectBox/selectBox.ts

```diff
@@ -5,56 +5,190 @@
+		const optionText = (<ISelectOptionItem>element).optionText;
+		const optionDisabled = (<ISelectOptionItem>element).optionDisabled;
+
+		data.optionText.textContent = optionText;
+		data.root.setAttribute('aria-label', nls.localize('selectAriaOption', "{0}", optionText));
+
+		// pseudo-select disabled option
+		if (optionDisabled) {
+			dom.addClass((<HTMLElement>data.root), 'option-disabled');
+		}
+	}
+
+	disposeTemplate(templateData: ISelectListTemplateData): void {
+		templateData.disposables = dispose(templateData.disposables);
+	}
+}
+
+export interface ISelectBoxStyles extends IListStyles {
 	selectBackground?: Color;
 	selectForeground?: Color;
 	selectBorder?: Color;
+	selectOptionsBorder?: Color;
+	selectOptionCheckedForeground?: Color;
+	selectOptionCheckedBackground?: Color;
+	selectOptionHoverForeground?: Color;
+	selectOptionHoverBackground?: Color;
+	selectOptionCheckedOutline?: Color;
+	selectOptionHoverOutline?: Color;
 }
 
 export const defaultStyles = {
 	selectBackground: Color.fromHex('#3C3C3C'),
 	selectForeground: Color.fromHex('#F0F0F0'),
-	selectBorder: Color.fromHex('#3C3C3C')
+	selectBorder: Color.fromHex('#3C3C3C'),
+	selectOptionsBorder: Color.fromHex('#3C3Ca0'),
+	selectOptionCheckedForeground: Color.fromHex('#ffffff'),
+	selectOptionCheckedBackground: Color.fromHex('#073655'),
+	selectOptionHoverForeground: Color.fromHex('#ffffff'),
+	selectOptionHoverBackground: Color.fromHex('#2A2D2E'),
+	selectOptionCheckedOutline: Color.fromHex('#F38518'),
+	selectOptionHoverOutline: Color.fromHex('#F38518')
 };
 
 export interface ISelectData {
 	selected: string;
 	index: number;
 }
 
-export class SelectBox extends Widget {
+export class SelectBox extends Widget implements IDelegate<ISelectOptionItem> {
+
+	private static SELECT_DROPDOWN_BOTTOM_MARGIN = 10;
 
+	private _useNativeSelect: boolean;
 	private selectElement: HTMLSelectElement;
 	private options: string[];
 	private selected: number;
+	private disabledOptionIndex: number;
 	private _onDidSelect: Emitter<ISelectData>;
 	private toDispose: IDisposable[];
-	private selectBackground: Color;
-	private selectForeground: Color;
-	private selectBorder: Color;
+	private styles: ISelectBoxStyles;
+	private listRenderer: SelectListRenderer;
+	private contextViewProvider: IContextViewProvider;
+	private selectDropDownContainer: HTMLElement;
+	private styleElement: HTMLStyleElement;
+	private selectList: List<ISelectOptionItem>;
+	private selectDropDownListContainer: HTMLElement;
+	private widthControlElement: HTMLElement;
+	// private widthControDivElement: HTMLElement;
 
-	constructor(options: string[], selected: number, styles: ISelectBoxStyles = deepClone(defaultStyles)) {
+	constructor(options: string[], selected: number, contextViewProvider: IContextViewProvider, styles: ISelectBoxStyles = deepClone(defaultStyles)) {
 		super();
 
+		// Hardcode native select switch based on platform
+		switch (process.platform) {
+			case 'darwin':
+				this._useNativeSelect = true;
```

Comment:
@cleidigh since this file is growing rather large because it contains both implementations of native and custom rendering I suggest to extract the two into files of their own and reference them from this class? E.g. `selectBoxNative.ts` and `selectBoxCustom.ts` which then would be used from `selectBox.ts` depending on the OS. I think that makes maintainability easier in the future.